walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,154 @@
+        TimeRange newTime = TimeRange.fromStartEnd(notAvailableTimes.get(j).end(), notAvailableTimes.get(j + 1).start(), false);
+        if (newTime.duration() >= request.duration()) {
+          availableTimes.add(newTime);
+        }
+      }
+    }
+    return availableTimes;                              
+  }
+
+  /**
+  * Takes in a collection of Events and a MeetingRequest and is expected to return
+  * a collection of TimeRanges in which the MeetingRequest can be satisfied given the
+  * the constraints posed by the attendees' events. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<String> attendeesWithoutEvents = new ArrayList<>(request.getAttendees());
+    int initialAttendeeCount = attendeesWithoutEvents.size();
+    ArrayList<String> optionalAttendeesWithoutEvents = new ArrayList<>(request.getOptionalAttendees());
+    int initialOptionalAttendeeCount = optionalAttendeesWithoutEvents.size();
+    
+    // Stores all TimeRanges of mandatory attendees with no repetitions.
+    Collection<TimeRange> timesRequiredAttendees = new HashSet<>();
+    // Stores all TimeRanges of optional attendees with no repetitions.
+    Collection<TimeRange> timesOptionalAttendees = new HashSet<>();
+    // Stores all TimeRanges in which mandatory attendees are busy with no overlaps.
+    ArrayList<TimeRange> notAvailableTimes = new ArrayList<>();
+    // Stores all TimeRanges in which optional attendees are busy with no overlaps.
+    ArrayList<TimeRange> notOptionalTimes = new ArrayList<>();
+    // Stores all TimeRanges in which mandatory attendees can meet.
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    // Stores all TimeRanges in which optional attendees can meet.
+    ArrayList<TimeRange> optionalCalendar = new ArrayList<>();
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // No events to conflict with.
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+      
+    ArrayList<String> toDeleteMan = new ArrayList<>();
+    ArrayList<String> toDeleteOpt = new ArrayList<>();
+      
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (attendeesWithoutEvents.contains(attendee)) {
+          // Adds TimeRange if any mandatory attendee attends this event.
+          timesRequiredAttendees.add(event.getWhen());
+          // Adds whenever a mandatory attendee appears in an event. 
+          toDeleteMan.add(attendee);
+        }
+        if (optionalAttendeesWithoutEvents.contains(attendee)) {
+          timesOptionalAttendees.add(event.getWhen());
+          toDeleteOpt.add(attendee);
+        }
+      }
+    }
+    // Deletes from our initial attendee counts those who appears on events so we can check 
```

Comment:
I don't think this comment explains why you've chosen to remove from the relevant lists here rather than in the loop above.

Reply:
Ahh, I think I see the reason now. If you changed nothing else then you'd end up missing events for individuals you've already removed from the attendeesWithoutEvents and optionalAttendeesWithoutEvents lists. The solution to that is to instead call the if statement with request.getAttendees() and request.getOptionalAttendees() (which is more intuitive anyway).